File: tests/loadtest_pure/sender.py
from geventhttpclient import Session
import argparse
import random
import gevent
import time


def random_in(a, b):
    return a + random.random() * (b - a)


def sender(
    user: int,
    req_per_sec: float,
    N: int,
    initialization_delay=(0, 0.5),
):
    gevent.sleep(random_in(*initialization_delay))
    s = Session()
    t0 = time.time()
    host = "http://127.0.0.1:8000"

    for n in range(N):
        s.get(
            url=host + "/send-hello",
            params={
                "user_id": user,
                "seq_id": n + 1,
            },
        )
        t = time.time()
        if (t - t0) < n / req_per_sec:
            gevent.sleep(n / req_per_sec - (t - t0))


if __name__ == "__main__":
    parser = argparse.ArgumentParser(description="Sender function arguments")
    parser.add_argument("-u", "--user", type=int, help="User ID")
    parser.add_argument(
        "-r",
        "--req_per_sec",
        type=float,
        default=3,
        help="Requests per second",
    )
    parser.add_argument(
        "-N",
        type=int,
        default=20,
        help="Requests per second",
    )
    args = parser.parse_args()
    # Convert the Namespace object to a dictionary and unpack it into keyword arguments
    sender(**vars(args))

File: tests/loadtest_go/sender.go
package sender

import (
	"fmt"
	"math/rand"
	"net/http"
	"time"
)

func randomIn(a, b float64) float64 {
	return a + rand.Float64()*(b-a)
}

func sender(user int, reqPerSec float64, n int, initializationDelay float64) {
	time.Sleep(time.Duration(randomIn(0, initializationDelay)) * time.Second)
	t0 := time.Now()
	host := "http://127.0.0.1:8000"
	for i := 0; i < n; i++ {
		resp, err := http.Get(fmt.Sprintf("%s/send-hello?user_id=%d&seq_id=%d", host, user, i+1))
		if err != nil {
			fmt.Println(err)
			return
		}
		defer resp.Body.Close()
		t := time.Now()
		if t.Sub(t0).Seconds() < float64(i)/reqPerSec {
			time.Sleep(time.Duration(float64(i)/reqPerSec-t.Sub(t0).Seconds()) * time.Second)
		}
	}
}
